File: src/app/core/services/pharmacie/medicament.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import {
  Medicament,
  MouvementRequest,
  StockStats,
} from '../../models/pharmacie/medicament.model';
import { Page } from '../../models/all/all.model';
import { ApiResponse } from '../../models/response/api-response.model';
import { CommonService } from '../common.services';
import { Entite } from '../../models/enums/enums.model';
import { environment } from '../../../../environments/environment';
import { Configuration } from '../../models/configuration/configuration.model';

@Injectable({ providedIn: 'root' })
export class MedicamentService {
  private http = inject(HttpClient);
  private commonService = inject(CommonService);

  private readonly API = `${environment.apiUrl}/pharmacie/medicaments`;

  findAll(
    page = 0,
    size = Configuration.pageSize,
    search = '',
    actif?: boolean,
  ): Observable<Page<Medicament>> {
    console.log('searchQuery dans le search :', search);
    let params = new HttpParams().set('page', page).set('size', size);
    if (search) params = params.set('q', search);
    if (actif !== undefined) params = params.set('actif', actif);
    console.log('Params HTTP:', params.toString());
    return this.http
      .get<ApiResponse<Page<Medicament>>>(this.API, { params })
      .pipe(
        map((r) => r.data),
        catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
      );
  }

  findById(id: string): Observable<Medicament> {
    return this.http.get<ApiResponse<Medicament>>(`${this.API}/${id}`).pipe(
      map((r) => r.data),
      catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
    );
  }

  create(req: Partial<Medicament>): Observable<Medicament> {
    return this.http.post<ApiResponse<Medicament>>(this.API, req).pipe(
      map((r) => r.data),
      catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
    );
  }

  edit(id: string, req: Partial<Medicament>): Observable<Medicament> {
    return this.http
      .put<ApiResponse<Medicament>>(`${this.API}/${id}`, req)
      .pipe(
        map((r) => r.data),
        catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
      );
  }

  delete(id: string): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.API}/${id}`).pipe(
      map(() => void 0),
      catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
    );
  }

  mouvement(req: MouvementRequest): Observable<Medicament> {
    return this.http
      .post<ApiResponse<Medicament>>(`${this.API}/mouvement`, req)
      .pipe(
        map((r) => r.data),
        catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
      );
  }

  getAlertes(): Observable<Medicament[]> {
    return this.http.get<ApiResponse<Medicament[]>>(`${this.API}/alertes`).pipe(
      map((r) => r.data),
      catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
    );
  }

  getRuptures(): Observable<Medicament[]> {
    return this.http
      .get<ApiResponse<Medicament[]>>(`${this.API}/ruptures`)
      .pipe(
        map((r) => r.data),
        catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
      );
  }

  getStats(): Observable<StockStats> {
    return this.http.get<ApiResponse<StockStats>>(`${this.API}/stats`).pipe(
      map((r) => r.data),
      catchError((e) => this.commonService.handleError(e, Entite.MEDICAMENT)),
    );
  }
}

File: pharmacie/src/main/java/it/solutions/services/trinity/pharmacie/services/MedicamentService.java
package it.solutions.services.trinity.pharmacie.services;


import it.solutions.services.trinity.contracts.dto.EmployeDto;
import it.solutions.services.trinity.contracts.dto.MedicamentDto;
import it.solutions.services.trinity.core.security.filters.JwtAuthFilter;
import it.solutions.services.trinity.core.shared.entities.User;
import it.solutions.services.trinity.core.shared.enums.Role;
import it.solutions.services.trinity.core.shared.enums.StatutConsultation;
import it.solutions.services.trinity.contracts.dto.ConsultationDto;
import it.solutions.services.trinity.core.shared.enums.StatutStock;
import it.solutions.services.trinity.pharmacie.dao.MedicamentDao;
import it.solutions.services.trinity.pharmacie.entities.Medicament;
import it.solutions.services.trinity.pharmacie.helpers.MedicamentHelper;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;



@Service
@RequiredArgsConstructor
public class MedicamentService {
    private static final Logger log = LoggerFactory.getLogger(MedicamentService.class);


    private static final Sort SORT_BY_NOM = Sort.by("nom").ascending();
    private final MedicamentDao dao;
    private final MedicamentHelper helper;


    public Page<MedicamentDto.Response> search(int page, int size, String search, Boolean actif) {
        Pageable pageable = PageRequest.of(page, size, SORT_BY_NOM);
        return dao.search(search, actif, pageable).map(helper::toResponse);
    }

    @Cacheable(value = "medicaments", key = "#id")
    public MedicamentDto.Response findById(UUID id) {
        return helper.toResponse(helper.getOrThrow(id));
    }

    @Transactional
    public MedicamentDto.Response create(MedicamentDto.Request req) {
        return helper.toResponse(dao.save(helper.builder(req)));
    }

    @Transactional
    @CacheEvict(value = "medicaments", key = "#id")
    public MedicamentDto.Response edit(UUID id, MedicamentDto.Request req) {

        Medicament medicament = helper.getOrThrow(id);
        helper.update(medicament ,req);
        return helper.toResponse(dao.save(medicament));


    }

    @Transactional
    @CacheEvict(value = "medicaments", key = "#id")
    public void delete(UUID id) {
        Medicament m = helper.getOrThrow(id);
        m.setActif(false); // soft delete
        dao.save(m);
    }

    // ── Mouvement de stock ────────────────────────────────
    @Transactional
    @CacheEvict(value = "medicaments", key = "#req.medicamentId")
    public MedicamentDto.Response mouvementStock(MedicamentDto.MouvementRequest req) {
        Medicament m = helper.getOrThrow(req.getMedicamentId());
        int nouveauStock = m.getStockActuel() + req.getQuantite();
        if (nouveauStock < 0) {
            throw new IllegalArgumentException(
                    "Stock insuffisant. Disponible : " + m.getStockActuel() +
                            ", demandé : " + Math.abs(req.getQuantite())
            );
        }
        m.setStockActuel(nouveauStock);
        Medicament saved = dao.save(m);
        log.info("Mouvement stock [{}] {} {} — {} (ref: {})",
                req.getMotif(), req.getQuantite() > 0 ? "+" : "",
                req.getQuantite(), m.getNom(), req.getReference());
        return helper.toResponse(saved);
    }

    // ── Alertes & stats ───────────────────────────────────
    public List<MedicamentDto.Response> getAlertes() {
        return dao.findEnAlerte().stream().map(helper::toResponse).toList();
    }

    public List<MedicamentDto.Response> getRuptures() {
        return dao.findEnRupture().stream().map(helper::toResponse).toList();
    }

    public List<MedicamentDto.Response> getExpirantBientot() {
        return dao.findExpirantAvant(LocalDate.now().plusDays(30))
                .stream().map(helper::toResponse).toList();
    }

    public MedicamentDto.StockStats getStats() {
        return MedicamentDto.StockStats.builder()
                .totalMedicaments(dao.countByActifTrue())
                .disponibles(dao.countDisponibles())
                .alertes(dao.countAlertes())
                .ruptures(dao.countRuptures())
                .expires(dao.countExpires())
                .expireBientot(dao.countExpirantBientot(LocalDate.now().plusDays(30)))
                .build();
    }

    // ── CRON : log des alertes chaque matin à 8h ─────────
    @Scheduled(cron = "0 0 8 * * *")
    public void verifierAlertes() {
        long alertes  = dao.countAlertes();
        long ruptures = dao.countRuptures();
        if (alertes > 0 || ruptures > 0) {
            log.warn("PHARMACIE — {} médicament(s) en alerte, {} en rupture", alertes, ruptures);
        }
    }
}
